Tests/AK/TestCharacterTypes.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibTest/TestCase.h>
+
+#include <AK/CharacterTypes.h>
+#include <AK/NumericLimits.h>
+#include <ctype.h>
+
+#define ASCII 0x80
+#define UNICODE 0x10FFFF + 100
+
+#define COMPARE_BOOLEAN_OUTPUT_OVER(range, func1, func2)               \
+    bool result1 = false;                                              \
+    bool result2 = false;                                              \
+    for (u32 i = 0; i < range; ++i) {                                  \
+        EXPECT_EQ(result1 = (func1(i) > 0), result2 = (func2(i) > 0)); \
+        if (result1 != result2)                                        \
+            dbgln("Function input value was {}.", i);                  \
+    }
+
+#define COMPARE_VALUE_OUTPUT_OVER(range, func1, func2)     \
+    bool result1 = false;                                  \
+    bool result2 = false;                                  \
+    for (u32 i = 0; i < range; ++i) {                      \
+        EXPECT_EQ(result1 = func1(i), result2 = func2(i)); \
+        if (result1 != result2)                            \
+            dbgln("Function input value was {}.", i);      \
+    }
+
+TEST_CASE(is_ascii)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(UNICODE, isascii, is_ascii);
+}
+
+TEST_CASE(is_ascii_alphanumeric)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII, isalnum, is_ascii_alphanumeric);
+}
+
+TEST_CASE(is_ascii_blank)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII, isblank, is_ascii_blank);
+}
+
+TEST_CASE(is_ascii_c0_control)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII - 1, iscntrl, is_ascii_c0_control);
+}
+
+TEST_CASE(is_ascii_control)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII, iscntrl, is_ascii_control);
+}
+
+TEST_CASE(is_ascii_digit)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII, isdigit, is_ascii_digit);
+}
+
+TEST_CASE(is_ascii_graphical)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII, isgraph, is_ascii_graphical);
+}
+
+TEST_CASE(is_ascii_hex_digit)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII, isxdigit, is_ascii_hex_digit);
+}
+
+TEST_CASE(is_ascii_lower_alpha)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII, islower, is_ascii_lower_alpha);
+}
+
+TEST_CASE(is_ascii_printable)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII, isprint, is_ascii_printable);
+}
+
+TEST_CASE(is_ascii_punctuation)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII, ispunct, is_ascii_punctuation);
+}
+
+TEST_CASE(is_ascii_space)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII, isspace, is_ascii_space);
+}
+
+TEST_CASE(is_ascii_upper_alpha)
+{
+    COMPARE_BOOLEAN_OUTPUT_OVER(ASCII, isupper, is_ascii_upper_alpha);
+}
+
+TEST_CASE(to_ascii_lowercase)
+{
+    COMPARE_VALUE_OUTPUT_OVER(UNICODE, tolower, to_ascii_lowercase);
+}
+
+TEST_CASE(to_ascii_uppercase)
+{
+    COMPARE_VALUE_OUTPUT_OVER(UNICODE, toupper, to_ascii_uppercase);
+}
```

Comment:
What about `parse_ascii_hex_digit(..)` and the other functions? Should we add tests for them as well?

Reply:
The problem I saw was how to test them? If you want to do it properly, you are just writing the same implementation a second time. The functions with equivalent functions in `ctype.h` can be compared to them, but to compare the others, you would need to use something like:
```cpp
for (u32 code_point = 0; code_point < 0x10FFF; ++code_point) {
    if (code_point >= 'A' && code_point <= 'Z')
        EXPECT_EQ(to_ascii_lowercase(code_point), code_point - 0x20);
    else
        EXPECT_EQ(to_ascii_lowercase(code_point), code_point);
}
```
This is literally just a second implementation of the same function and in my opinion worth about as much as:
```cpp
EXPECT_EQ(to_ascii_lowercase(code_point), to_ascii_lowercase(code_point));
```

Reply:
Sure, maybe it makes sense to validate a few hard coded values? Up to you... Just asking questions 😃